### PR TITLE
Internal block api unit testing

### PR DESCRIPTION
**Pull request by ssd04**
Unit testing on several parts regarding internal block headers exposure:
- api
- facade
- status handler ratings metrics
- node internal blocks

### CODE REVIEW

File: statusHandler/statusMetricsProvider.go [hunk 1]
```diff
@@ -255,7 +255,7 @@ func (sm *statusMetrics) RatingsMetrics() map[string]interface{} {
 	ratingsMetrics[common.MetricRatingsGeneralSignedBlocksThreshold] = sm.loadStringMetric(common.MetricRatingsGeneralSignedBlocksThreshold)
 
 	numSelectionChances := sm.loadUint64Metric(common.MetricRatingsGeneralSelectionChances + "_count")
-	selectionChances := make([]interface{}, 0)
+	selectionChances := make([]map[string]interface{}, 0)
```
Comment (bogdan-rosianu) [inline statusHandler/statusMetricsProvider.go:258]: the items within the map seem only to be `uint64`. therefore, we could change the map from `map[string]interface{}` to `map[string]uint32`

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: statusHandler/statusMetricsProvider_test.go [hunk 2]
```diff
@@ -250,3 +250,92 @@ func TestStatusMetrics_EnableEpochMetrics(t *testing.T) {
...
+		},
+		{
+			"MaxThreshold":  1000,
+			"ChancePercent": 10,
+		},
+	}
+
+	for i, selectionChance := range selectionChances {
+		maxThresholdStr := fmt.Sprintf("%s%d%s", common.MetricRatingsGeneralSelectionChances, i, common.SelectionChancesMaxThresholdSuffix)
+		sm.SetUInt64Value(maxThresholdStr, uint64(selectionChance["MaxThreshold"]))
```
Comment (bogdan-rosianu) [inline statusHandler/statusMetricsProvider_test.go:277]: Redundant type conversion

File: statusHandler/statusMetricsProvider_test.go [hunk 3]
```diff
@@ -250,3 +250,92 @@ func TestStatusMetrics_EnableEpochMetrics(t *testing.T) {
...
+		chancePercentStr := fmt.Sprintf("%s%d%s", common.MetricRatingsGeneralSelectionChances, i, common.SelectionChancesChancePercentSuffix)
+		sm.SetUInt64Value(chancePercentStr, uint64(selectionChance["ChancePercent"]))
```
Comment (bogdan-rosianu) [inline statusHandler/statusMetricsProvider_test.go:279]: Redundant type conversion

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: api/groups/internalGroup_test.go [hunk 4]
```diff
@@ -105,23 +189,177 @@ func TestGetRawMetaBlockByNonce_ShouldWork(t *testing.T) {
...
+			return expectedOutput, nil
+		},
+	}
+
+	blockGroup, err := groups.NewInternalBlockGroup(&facade)
+	require.NoError(t, err)
+
+	ws := startWebServer(blockGroup, "internal", getInternalBlockRoutesConfig())
+
+	req, _ := http.NewRequest("GET", "/internal/raw/metablock/by-hash/15", nil)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply api/groups/internalGroup_test.go:362]: done

File: api/groups/internalGroup_test.go [hunk 5]
```diff
@@ -105,23 +189,177 @@ func TestGetRawMetaBlockByNonce_ShouldWork(t *testing.T) {
...
+	assert.True(t, strings.Contains(response.Error, expectedErr.Error()))
+}
+
+func TestGetRawMetaBlockByHash_ShouldWork(t *testing.T) {
+	t.Parallel()
+
+	expectedOutput := bytes.Repeat([]byte("1"), 10)
+
+	facade := mock.FacadeStub{
+		GetInternalMetaBlockByHashCalled: func(_ common.OutportFormat, _ string) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply api/groups/internalGroup_test.go:352]: done

File: node/blockAPI/internalBlock_test.go [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+	blk, err := internalBlockProcessor.GetInternalShardBlockByRound(common.Proto, round)
+	assert.Nil(t, err)
+	assert.Equal(t, headerBytes, blk)
+}
+
+func preapreShardBlockProcessor(nonce uint64, round uint64, headerHash []byte) (*internalBlockProcessor, []byte) {
+	storerMock := mock.NewStorerMock()
+	uint64Converter := mock.NewNonceHashConverterMock()
+
+	internalBlockProcessor := createMockInternalBlockProcessor(
```
[the remark this answers is not in the document]
Comment (ssd04) [reply node/blockAPI/internalBlock_test.go:280]: done

File: node/node.go [hunk 7]
```diff
@@ -1404,7 +1404,8 @@ func (n *Node) getKeyBytes(key string) ([]byte, error) {
 	return hex.DecodeString(key)
 }
 
-func (n *Node) createInternalBlockProcessor() error {
+// CreateInternalBlockProcessor creates the block processor for handling internal block header
+func (n *Node) CreateInternalBlockProcessor() error {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply node/node.go:1408]: right, exported function in `export_test.go`

File: statusHandler/statusMetricsProvider.go (hunk 1 above)
Comment (ssd04) [reply statusHandler/statusMetricsProvider.go:258]: right :+1:

File: statusHandler/statusMetricsProvider_test.go (hunk 2 above)
Comment (ssd04) [reply statusHandler/statusMetricsProvider_test.go:277]: done

File: statusHandler/statusMetricsProvider_test.go (hunk 3 above)
Comment (ssd04) [reply statusHandler/statusMetricsProvider_test.go:279]: done

File: node/nodeInternalBlocks_test.go [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+	n, _ := node.NewNode(
+		node.WithCoreComponents(coreComponentsMock),
+		node.WithProcessComponents(processComponentsMock),
+		node.WithDataComponents(dataComponentsMock),
+	)
+
+	err := n.CreateInternalBlockProcessor()
+	require.Nil(t, err)
+
+	blk, err := n.GetInternalShardBlockByHash(common.Internal, "wronghashformat")
```
[the remark this answers is not in the document]
Comment (ssd04) [reply node/nodeInternalBlocks_test.go]: done :+1:

**Review by bogdan-rosianu — DISMISSED** [review]

File: node/nodeInternalBlocks_test.go [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+	dataComponentsMock.Store = nil
+
+	n, _ := node.NewNode(
+		node.WithCoreComponents(coreComponentsMock),
+		node.WithProcessComponents(processComponentsMock),
+		node.WithDataComponents(dataComponentsMock),
+	)
+
+	err := n.CreateInternalBlockProcessor()
+	assert.Error(t, err)
```
Comment (AdoAdoAdo) [inline node/nodeInternalBlocks_test.go:40]: is it possible to check against the specific error?
check also the other tests

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: api/groups/internalGroup_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply api/groups/internalGroup_test.go:362]: right :+1: , I changed all the other except this one :)

**Review by AdoAdoAdo — APPROVED** [review]

File: node/blockAPI/error.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package blockAPI
+
+import "errors"
+
+// ErrInvalidOutputFormat signals that the outport format type is not valid
+var ErrInvalidOutputFormat = errors.New("the outport format type is invalid")
```
Comment (AdoAdoAdo) [inline node/blockAPI/error.go:6]: typo outport -> output

File: node/nodeInternalBlocks_test.go [hunk 11]
```diff
@@ -36,8 +38,10 @@ func TestCreateInternalBlockProcessor_NilStoreShouldErr(t *testing.T) {
 		node.WithDataComponents(dataComponentsMock),
 	)
 
+	expectedErr := errors.New(
+		"error creating transaction status computer " + txstatus.ErrNiStorageService.Error())
 	err := n.CreateInternalBlockProcessor()
-	assert.Error(t, err)
+	assert.Equal(t, err, expectedErr)
```
Comment (AdoAdoAdo) [inline node/nodeInternalBlocks_test.go:44]: you can check with something like:
assert.True(t, errors.Is(err, data.ErrNiStorageService))
then you don't need the expectedErr creation

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

File: node/blockAPI/error.go (hunk 10 above)
Comment (ssd04) [reply node/blockAPI/error.go]: right. thanks

File: node/nodeInternalBlocks_test.go (hunk 11 above)
Comment (ssd04) [reply node/nodeInternalBlocks_test.go]: done

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**